### PR TITLE
add Units::getPhysicalDescription

### PR DESCRIPTION
**Pull request by nornagon**
This returns the physical unit description seen on the unit screen, e.g.

> His very long sideburns are braided.  His very long moustache is arranged in double braids.  His very long beard is arranged in double braids.  His hair is clean-shaven.  He has a scratchy voice.  His free-lobed ears are splayed out.  His nose is narrow.  His lips are very thin.  His nose bridge is slightly convex.  His dark peach skin is slightly wrinkled.  His emerald eyes have slightly thin irises.  His hair is ecru with a touch of gray.

There's no easy way to reproduce this string directly from the body information, as there's a substantial amount of logic that goes into deciding what appears in this string and in what order.

Requires a new symbol in symbols.txt, which I've currently only found for macos x64 (0x100cbb890), but I'll find it on other platforms if this is plausibly mergeable! I found it by setting a memory-read breakpoint on a constant string that appears in that description (e.g. "incredibly short"), then jumping up the stack trace until I found this function.

### CODE REVIEW

**Comment by quietust** [comment]
In case it wasn't already clear, this new symbol is for *code*, not data. The function appears to be a (non-virtual) method of the "unitst" class, and it takes 1 parameter aside from `this` (to take the string).

The prototype you provided will **not** work in the 32-bit Windows binary (whose address happens to be 0xDB87F0) - you must specify a calling convention of `__thiscall` to ensure that the first parameter ("this") is passed in the `ecx` register instead of on the stack.

**Comment by nornagon** [comment]
Ah, good find. Thanks. I tested this on macos and windows x64 but not on any 32bit platforms.

I think this can be made to work using something like:

```cpp
class dummy_unitst {
  public:
    void physical_description(std::string *out_str);
};

// ...
std::string str;
reinterpret_cast<dummy_unitst*>(unit)->physical_description(&str);
```

I'll experiment with this soon and update the PR.

**Comment by nornagon** [comment]
... oh wait, can we just write `__thiscall` on windows and it'll work?

**Comment by lethosor** [comment]
I believe so - exportmaps and old versions of TWBT might have examples

**Comment by nornagon** [comment]
Still just experimenting here, I'll try this out for real when I'm back at my windows machine.

**Comment by lethosor** [comment]
Looks about right, but you might want to verify that _WIN32 isn't also defined on 64-bit Windows (I think WIN32 is, at least)

**Comment by nornagon** [comment]
Per comment in the commit, __thiscall does nothing on x64 anyway.

File: library/modules/Units.cpp [hunk 1]
```diff
@@ -541,6 +541,23 @@ string Units::getRaceName(df::unit* unit)
     return getRaceNameById(unit->race);
 }
 
+void df_unit_physical_description(df::unit* unit, string* out_str)
+{
+    static auto* const fn =
+        reinterpret_cast<void(THISCALL *)(df::unit*, string*)>(
+            Core::getInstance().vinfo->getAddress("unit_physical_description"));
+    CHECK_NULL_POINTER(fn);
```
Comment (lethosor) [inline library/modules/Units.cpp]: I would actually keep the do-nothing behavior that you had earlier if fn is null - when global addresses are missing, most other functions also fall back to some default behavior. Setting out_str to an empty string is also an option.

Comment (nornagon) [reply library/modules/Units.cpp]: done 👍

**Comment by nornagon** [comment]
What do you think of calling the symbol something like
`unit::getPhysicalDescription(std::string*)`?
-- 

j

**Comment by lethosor** [comment]
Special characters in symbol names might break things - they won't be valid code if they end up in generated code, and there might be some tools we have that can't handle that, so I'd avoid it

**Comment by nornagon** [comment]
One more potential alternative, since there might in future be other function pointers in the symbols that aren't getters: `unit_get_physical_description`? i.e. put the class name first. Or perhaps even `unit__get_physical_description`, having `__` stand in for `::`?

**Comment by lethosor** [comment]
Good point. I like `unit_get_physical_description` - it's consistent with some other "groups" of global addresses, like `ui_`, `standing_orders_`, etc.

**Pull request merged**